### PR TITLE
Feature/migrations setup

### PR DESCRIPTION
**Pull request by stdroniv**

### CODE REVIEW

File: migrations/000001_create_authors_table.up.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE IF NOT EXISTS authors (
+    id SERIAL PRIMARY KEY,
+    name VARCHAR(255) NOT NULL UNIQUE,
```
Comment (obaibula) [inline migrations/000001_create_authors_table.up.sql]: We can use text

File: migrations/000001_create_authors_table.up.sql [hunk 2]
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE IF NOT EXISTS authors (
+    id SERIAL PRIMARY KEY,
```
Comment (obaibula) [inline migrations/000001_create_authors_table.up.sql]: Bigserial would be better

File: migrations/000002_create_categories_table.up.sql [hunk 3]
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE IF NOT EXISTS categories (
+    id SERIAL PRIMARY KEY,
```
Comment (obaibula) [inline migrations/000002_create_categories_table.up.sql]: Bigserial would be better

File: migrations/000003_create_books_table.up.sql [hunk 4]
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE IF NOT EXISTS books (
+    id SERIAL PRIMARY KEY,
```
Comment (obaibula) [inline migrations/000003_create_books_table.up.sql]: Bigserial would be better

File: migrations/000003_create_books_table.up.sql [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    title VARCHAR(255) NOT NULL,
```
Comment (obaibula) [inline migrations/000003_create_books_table.up.sql]: Could be text

File: migrations/000003_create_books_table.up.sql [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    isbn VARCHAR(13) UNIQUE NOT NULL,
+    author_id INT NOT NULL REFERENCES authors(id) ON DELETE CASCADE,
+    publisher VARCHAR(255),
```
Comment (obaibula) [inline migrations/000003_create_books_table.up.sql]: Could be text and Should have default value

File: migrations/000003_create_books_table.up.sql [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    price DECIMAL(10, 2) NOT NULL,
+    stock INT DEFAULT 0,
+    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
+);
+
+CREATE INDEX IF NOT EXISTS idx_books_isbn ON books(isbn);
```
Comment (obaibula) [inline migrations/000003_create_books_table.up.sql]: Redundant, on unique constraints Postgres creates indecies automatically

File: migrations/000001_create_authors_table.up.sql [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+    name VARCHAR(255) NOT NULL UNIQUE,
+    bio TEXT,
+    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
+);
+
+CREATE INDEX IF NOT EXISTS idx_authors_name ON authors(name);
```
Comment (obaibula) [inline migrations/000001_create_authors_table.up.sql]: Redundant, on unique constraints Postgres creates indecies automatically

File: migrations/000001_create_authors_table.up.sql [hunk 9]
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE IF NOT EXISTS authors (
+    id SERIAL PRIMARY KEY,
+    name VARCHAR(255) NOT NULL UNIQUE,
+    bio TEXT,
```
Comment (obaibula) [inline migrations/000001_create_authors_table.up.sql]: Should have default value

File: migrations/000001_create_authors_table.down.sql [hunk 10]
```diff
@@ -0,0 +1,3 @@
+DROP TABLE IF EXISTS authors;
+
+DROP INDEX IF EXISTS idx_authors_name;
```
Comment (obaibula) [inline migrations/000001_create_authors_table.down.sql]: Redundant, see up file

File: migrations/000002_create_categories_table.up.sql [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+    name VARCHAR(255) NOT NULL UNIQUE,
+    description TEXT,
+    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
+);
+
+CREATE INDEX IF NOT EXISTS idx_categories_name ON categories(name);
```
Comment (obaibula) [inline migrations/000002_create_categories_table.up.sql]: Redundant, on unique constraints Postgres creates indecies automatically

File: migrations/000002_create_categories_table.down.sql [hunk 12]
```diff
@@ -0,0 +1,3 @@
+DROP TABLE IF EXISTS categories;
+
+DROP INDEX IF EXISTS idx_categories_name;
```
Comment (obaibula) [inline migrations/000002_create_categories_table.down.sql]: Redundant

File: migrations/000002_create_categories_table.up.sql [hunk 13]
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE IF NOT EXISTS categories (
+    id SERIAL PRIMARY KEY,
+    name VARCHAR(255) NOT NULL UNIQUE,
```
Comment (obaibula) [inline migrations/000002_create_categories_table.up.sql]: Could be text

File: migrations/000003_create_books_table.down.sql [hunk 14]
```diff
@@ -0,0 +1,4 @@
+DROP TABLE IF EXISTS books;
+
+DROP INDEX IF EXISTS idx_books_isbn;
```
Comment (obaibula) [inline migrations/000003_create_books_table.down.sql]: Redundant

File: migrations/000002_create_categories_table.up.sql [hunk 15]
```diff
@@ -0,0 +1,9 @@
...
+    description TEXT,
```
Comment (obaibula) [inline migrations/000002_create_categories_table.up.sql]: Should have default value

File: migrations/000004_create_book_categories_table.up.sql [hunk 16]
```diff
@@ -0,0 +1,5 @@
+CREATE TABLE IF NOT EXISTS book_categories (
+    book_id INT NOT NULL REFERENCES books(id) ON DELETE CASCADE,
```
Comment (obaibula) [inline migrations/000004_create_book_categories_table.up.sql]: bigint

File: migrations/000004_create_book_categories_table.up.sql [hunk 17]
```diff
@@ -0,0 +1,5 @@
...
+    category_id INT NOT NULL REFERENCES categories(id) ON DELETE CASCADE,
```
Comment (obaibula) [inline migrations/000004_create_book_categories_table.up.sql]: bigint

File: migrations/000003_create_books_table.up.sql [hunk 18]
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE IF NOT EXISTS books (
+    id SERIAL PRIMARY KEY,
+    title VARCHAR(255) NOT NULL,
+    isbn VARCHAR(13) UNIQUE NOT NULL,
+    author_id INT NOT NULL REFERENCES authors(id) ON DELETE CASCADE,
```
Comment (obaibula) [inline migrations/000003_create_books_table.up.sql]: bigint

File: migrations/000003_create_books_table.up.sql [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+    publisher VARCHAR(255),
+    price DECIMAL(10, 2) NOT NULL,
+    stock INT DEFAULT 0,
+    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
+);
+
+CREATE INDEX IF NOT EXISTS idx_books_isbn ON books(isbn);
+CREATE INDEX IF NOT EXISTS idx_books_title ON books(title);
```
Comment (obaibula) [inline migrations/000003_create_books_table.up.sql]: It's better to use gin index with pg_vector, to fire full text search in future.

**Review by obaibula — CHANGES_REQUESTED** [review]
A lot of comments, but all about the same.

File: migrations/000001_create_authors_table.up.sql (hunk 2 above)
Comment (stdroniv) [reply migrations/000001_create_authors_table.up.sql]: Why exactly? Are you sure that range from 1 to 2billion wouldn't be enough for authors table? I'm not sure we have that much of authors

File: migrations/000003_create_books_table.up.sql (hunk 4 above)
Comment (stdroniv) [reply migrations/000003_create_books_table.up.sql]: Same for books

File: migrations/000004_create_book_categories_table.up.sql (hunk 16 above)
Comment (stdroniv) [reply migrations/000004_create_book_categories_table.up.sql]: int limitation is 2147483647, are you sure we would have more items?

File: migrations/000004_create_book_categories_table.up.sql (hunk 17 above)
Comment (stdroniv) [reply migrations/000004_create_book_categories_table.up.sql]: int limitation is 2147483647, are you sure we would have more items?

File: migrations/000003_create_books_table.up.sql (hunk 18 above)
Comment (stdroniv) [reply migrations/000003_create_books_table.up.sql]: int limitation is 2147483647, are you sure we would have more items?

Comment (obaibula) [reply migrations/000003_create_books_table.up.sql]: We would have less than 100 books, should we yse smallint?

File: migrations/000002_create_categories_table.up.sql [hunk 20]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE IF NOT EXISTS categories (
+    id bigint GENERATED ALWAYS AS IDENTITY,
```
Comment (obaibula) [inline migrations/000002_create_categories_table.up.sql]: Can wecchange bigint to uppercase for consistency?))

**Review by obaibula — APPROVED** [review]

**Pull request merged**